source/fe/mapping_q_generic.cc

```diff
@@ -108,21 +108,23 @@ namespace internal
         const double y2 = vertices[2](1);
         const double y3 = vertices[3](1);
 
-        const long double a = (x1 - x3)*(y0 - y2) - (x0 - x2)*(y1 - y3);
-        const long double b = -(x0 - x1 - x2 + x3)*y + (x - 2*x1 + x3)*y0 - (x - 2*x0 + x2)*y1
-                              - (x - x1)*y2 + (x - x0)*y3;
-        const long double c = (x0 - x1)*y - (x - x1)*y0 + (x - x0)*y1;
+        const double a = (x1 - x3)*(y0 - y2) - (x0 - x2)*(y1 - y3);
+        const double b = -(x0 - x1 - x2 + x3)*y + (x - 2*x1 + x3)*y0 - (x - 2*x0 + x2)*y1
+                         - (x - x1)*y2 + (x - x0)*y3;
+        const double c = (x0 - x1)*y - (x - x1)*y0 + (x - x0)*y1;
 
-        const long double discriminant = b*b - 4*a*c;
+        const double discriminant = b*b - 4*a*c;
         // exit if the point is not in the cell (this is the only case where the
         // discriminant is negative)
         AssertThrow (discriminant > 0.0,
                      (typename Mapping<spacedim,spacedim>::ExcTransformationFailed()));
 
-        long double eta1;
-        long double eta2;
-        // special case #1: if a is zero, then use the linear formula
-        if (a == 0.0 && b != 0.0)
+        double eta1;
+        double eta2;
+        const double sqrt_discriminant = std::sqrt(discriminant);
+        // special case #1: if a is near-zero to make the discriminant exactly
+        // equal b, then use the linear formula
+        if (b != 0.0 && std::abs(b) == sqrt_discriminant)
```

Comment:
Would that ever happen? I'm trying to think of a case where, say, `b==1`. If `a*c < 1e-16`, then the discriminant is indeed `b*b`. But is then `std::abs(b) == sqrt_discriminant`? Wouldn't it be better to test whether `b*b == discriminant`? I would expect that to be a more robust formula than to test that `std::abs(b) == std::sqrt(b*b)`.

Reply:
The new test case shows that we indeed go into this branch with the current check. Why do you think that testing for the square of the expression is more robust? I'm adding exactly the check for exactly the expression that appears five lines down in the division, `-b-sqrt_discriminant` and `-b+sqrt_discriminant`. Instead of checking whether either of those is zero, I only checked the two contributions with `abs`. I can change that if you want, but I don't see how the alternative would be more robust, since it involves additional rounding as compared to my check.

Reply:
No, it's fine. I'm just surprised that this case ever happens because I would have expected that `std::abs(b) != std::sqrt(b*b)` for most numbers `b`, but that `b*b == discriminant` can happen more frequently.

But since you want to catch one particular case, I guess this is ok with me.